File: app/src/main/java/com/rascarlo/aurdroid/adapters/SearchResultAdapter.java
/*
 *     Copyright (C) rascarlo  rascarlo@gmail.com
 *
 *     This program is free software: you can redistribute it and/or modify
 *     it under the terms of the GNU General Public License as published by
 *     the Free Software Foundation, either version 3 of the License, or
 *     (at your option) any later version.
 *
 *     This program is distributed in the hope that it will be useful,
 *     but WITHOUT ANY WARRANTY; without even the implied warranty of
 *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *     GNU General Public License for more details.
 *
 *     You should have received a copy of the GNU General Public License
 *     along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

package com.rascarlo.aurdroid.adapters;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.ListAdapter;
import androidx.recyclerview.widget.DiffUtil;
import android.text.TextUtils;
import android.view.LayoutInflater;
import android.view.ViewGroup;

import com.rascarlo.aurdroid.api.model.SearchResult;
import com.rascarlo.aurdroid.callbacks.SearchResultAdapterCallback;
import com.rascarlo.aurdroid.databinding.SearchResultItemBinding;
import com.rascarlo.aurdroid.viewholders.SearchResultAdapterViewHolder;

public class SearchResultAdapter extends ListAdapter<SearchResult, SearchResultAdapterViewHolder> {
    private final SearchResultAdapterCallback searchResultAdapterCallback;

    public SearchResultAdapter(SearchResultAdapterCallback searchResultAdapterCallback) {
        super(DIFF_CALLBACK);
        this.searchResultAdapterCallback = searchResultAdapterCallback;
    }

    private static final DiffUtil.ItemCallback<SearchResult> DIFF_CALLBACK = new DiffUtil.ItemCallback<SearchResult>() {
        @Override
        public boolean areItemsTheSame(@NonNull SearchResult searchResult, @NonNull SearchResult t1) {
            return TextUtils.equals(searchResult.getiD(), t1.getiD());
        }

        @Override
        public boolean areContentsTheSame(@NonNull SearchResult searchResult, @NonNull SearchResult t1) {
            return searchResult.equals(t1);
        }
    };

    @NonNull
    @Override
    public SearchResultAdapterViewHolder onCreateViewHolder(@NonNull ViewGroup viewGroup, int i) {
        SearchResultItemBinding searchResultItemBinding = SearchResultItemBinding.inflate(LayoutInflater.from(viewGroup.getContext()), viewGroup, false);
        searchResultItemBinding.setSearchResultAdapterCallback(searchResultAdapterCallback);
        return new SearchResultAdapterViewHolder(searchResultItemBinding);
    }

    @Override
    public void onBindViewHolder(@NonNull SearchResultAdapterViewHolder searchResultAdapterViewHolder, int i) {
        if (getItem(i) != null) {
            SearchResult searchResult = getItem(i);
            searchResultAdapterViewHolder.bindResult(searchResult);
        }
    }
}

File: app/src/main/java/com/rascarlo/aurdroid/searchResult/SearchResultAdapter.kt
package com.rascarlo.aurdroid.searchResult

import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.ListAdapter
import androidx.recyclerview.widget.RecyclerView
import com.rascarlo.aurdroid.databinding.SearchResultItemBinding
import com.rascarlo.aurdroid.network.SearchResult

/**
 * list adapter for [SearchResultFragment]
 */
class SearchResultAdapter(private val onClickListener: OnClickListener) :
    ListAdapter<SearchResult, SearchResultAdapter.SearchResultViewHolder>(DiffCallback) {

    /**
     * diff util callback for [SearchResult]
     * wrapper into a companion object
     * use [SearchResult] to compare if [areItemsTheSame]
     * use [SearchResult.id] to compare if [areContentsTheSame]
     */
    private companion object DiffCallback : DiffUtil.ItemCallback<SearchResult>() {
        override fun areItemsTheSame(oldItem: SearchResult, newItem: SearchResult): Boolean {
            return oldItem == newItem
        }

        override fun areContentsTheSame(oldItem: SearchResult, newItem: SearchResult): Boolean {
            return oldItem.id == newItem.id
        }
    }

    // view holder
    class SearchResultViewHolder(private var binding: SearchResultItemBinding) :
        RecyclerView.ViewHolder(binding.root) {
        fun bind(searchResult: SearchResult) {
            binding.searchResult = searchResult
            binding.executePendingBindings()
        }
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): SearchResultViewHolder {
        val layoutInflater = LayoutInflater.from(parent.context)
        return SearchResultViewHolder(
            SearchResultItemBinding.inflate(layoutInflater, parent, false)
        )
    }

    override fun onBindViewHolder(holder: SearchResultViewHolder, position: Int) {
        val searchResult = getItem(position)
        holder.bind(searchResult)
        holder.itemView.setOnClickListener { onClickListener.onClick(searchResult) }
    }

    class OnClickListener(val clickListener: (searchResult: SearchResult) -> Unit) {
        fun onClick(searchResult: SearchResult) = clickListener(searchResult)
    }
}